osquery/events/windows/etw/etw_publisher_processes.cpp

```diff
@@ -0,0 +1,552 @@
+
+#include <osquery/core/flags.h>
+#include <osquery/events/windows/etw/etw_publisher_processes.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/map_take.h>
+
+namespace osquery {
+
+FLAG(bool,
+     enable_etw_process_events,
+     false,
+     "Enables the etw_process_events publisher");
+
+// ETW Event publisher registration into the Osquery pub-sub framework
+REGISTER_ETW_PUBLISHER(EtwPublisherProcesses, kEtwProcessPublisherName.c_str());
+
+// Publisher constructor
+EtwPublisherProcesses::EtwPublisherProcesses()
+    : EtwPublisherBase(kEtwProcessPublisherName) {
+  initializeHardVolumeConversions();
+};
+
+// There are multiple ETW providers being setup here. Events arriving from
+// these providers will be aggregated in the post-processing phase.
+Status EtwPublisherProcesses::setUp() {
+  if (!FLAGS_enable_etw_process_events) {
+    return Status::failure(kEtwProcessPublisherName +
+                           " publisher disabled via configuration.");
+  }
+  // ETW Initialization logic
+  const EtwProviderConfig::EtwBitmask processStartStopKeyword = 0x10;
+
+  // Userspace ETW Provider configuration
+  EtwProviderConfig userEtwProviderConfig;
+  userEtwProviderConfig.setName("Microsoft-Windows-Kernel-Process");
+  userEtwProviderConfig.setAnyBitmask(processStartStopKeyword);
+  userEtwProviderConfig.setPreProcessor(getPreProcessorCallback());
+  userEtwProviderConfig.setPostProcessor(getPostProcessorCallback());
+  userEtwProviderConfig.addEventTypeToHandle(EtwEventType::ProcessStart);
+  userEtwProviderConfig.addEventTypeToHandle(EtwEventType::ProcessStop);
+
+  // Adding the provider to the ETW Engine
+  Status userProviderAddStatus = EtwEngine().addProvider(userEtwProviderConfig);
+  if (!userProviderAddStatus.ok()) {
+    return userProviderAddStatus;
+  }
+
+  // Kernelspace ETW Provider configuration
+  EtwProviderConfig kernelProviderCfg;
+  kernelProviderCfg.setKernelProviderType(
+      EtwProviderConfig::EtwKernelProviderType::Process);
+  kernelProviderCfg.setPreProcessor(getPreProcessorCallback());
+  kernelProviderCfg.setPostProcessor(getPostProcessorCallback());
+  kernelProviderCfg.addEventTypeToHandle(EtwEventType::ProcessStart);
+  kernelProviderCfg.addEventTypeToHandle(EtwEventType::ProcessStop);
+
+  // Adding the provider to the ETW Engine
+  Status kernelProviderAddStatus = EtwEngine().addProvider(kernelProviderCfg);
+  if (!kernelProviderAddStatus.ok()) {
+    return kernelProviderAddStatus;
+  }
+
+  return Status::success();
+}
+
+// Callback to perform pre-processing logic
+void EtwPublisherProcesses::providerPreProcessor(
+    const EVENT_RECORD& rawEvent, const krabs::trace_context& traceCtx) {
+  // Helper accessors for userspace events
+  const EVENT_HEADER& eventHeader = rawEvent.EventHeader;
+  const unsigned int eventVersion = eventHeader.EventDescriptor.Version;
+
+  // Checking if new event is a supported one
+  if (!isSupportedEvent(eventHeader)) {
+    return;
```

Comment:
I'm not sure how unusual this is, but maybe should log here?

Reply:
This callback is the entry point for events arriving from the ETW providers. This could end up producing a very verbose logging output, so I'll try to avoid adding logging here for the time being